### PR TITLE
Added `Option` alternative to `indexOf` methods

### PR DESCRIPTION
**Pull request by paplorinc**
Fixes https://github.com/javaslang/javaslang/issues/1278

### CODE REVIEW

File: javaslang/src/main/java/javaslang/collection/Seq.java [hunk 1]
```diff
@@ -337,14 +365,36 @@ private boolean checkPrefix(Seq<T> t, Seq<T> prefix) {
...
+     */
+    default Option<Integer> indexOfSliceOption(Iterable<? extends T> that, int from) {
+        return Collections.indexOption(indexOfSlice(that, from));
+    }
+
+    /**
      * Finds index of first element satisfying some predicate.
      *
-     * @param p the predicate used to test elements.
+     * @param predicate the predicate used to test elements.
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Seq.java]: renamed for consistency

File: javaslang/src/main/java/javaslang/collection/IndexedSeq.java [hunk 2]
```diff
@@ -366,6 +366,7 @@ default boolean startsWith(Iterable<? extends T> that, int offset) {
      * the return value will be &gt;= 0 if and only if the element is found.
      * @throws ClassCastException if T cannot be cast to {@code Comparable<? super T>}
      */
+    @Override
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/IndexedSeq.java]: done in separate commit

File: javaslang/src/main/java/javaslang/control/Option.java [hunk 3]
```diff
@@ -138,6 +133,18 @@
...
 
     /**
+     * Creates {@code Some} of value if condition is true, or {@code None} in other case
+     *
+     * @param <T>       type of the optional value
+     * @param condition A boolean value
+     * @param value     An optional value, may be {@code null}
+     * @return return {@code Some} of value if condition is true, or {@code None} in other case
+     */
+    static <T> Option<T> when(boolean condition, T value) {
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/control/Option.java]: done in separate commit

File: javaslang/src/test/java/javaslang/collection/CharSeqTest.java [hunk 4]
```diff
@@ -2613,27 +2654,27 @@ public void shouldSplitAtInclusivePredicateNotFound() {
     // -- removeAt(index)
 
     @Test(expected = IndexOutOfBoundsException.class)
-    public void shouldRemoveIndxAtNil() {
+    public void shouldRemoveIndexAtNil() {
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/CharSeqTest.java]: unrelated typo

File: javaslang/src/test/java/javaslang/collection/AbstractSeqTest.java [hunk 5]
```diff
@@ -338,87 +336,119 @@ public void shouldGetLastElement() {
     @Test
     public void shouldNotFindIndexOfElementWhenSeqIsEmpty() {
         assertThat(empty().indexOf(1)).isEqualTo(-1);
+
+        assertThat(empty().indexOfOption(1)).isEqualTo(Option.none());
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/AbstractSeqTest.java]: we could test only the `Option` alternatives, since they're only delegating. Your choice.

File: javaslang/src/main/java/javaslang/collection/Seq.java (hunk 1 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Seq.java]: good!

File: javaslang/src/main/java/javaslang/collection/CharSeq.java [hunk 6]
```diff
@@ -1605,6 +1605,16 @@ public int indexOf(int ch) {
...
 
     /**
+     * Returns the index of the first occurrence of the given element as an {@code Option}.
+     *
+     * @param ch a character (Unicode code point).
+     * @return {@code Some(index)} or {@code None} if not found.
+     */
+    Option<Integer> indexOfOption(int ch) {
+        return Collections.indexOption(indexOf(ch));
+    }
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/CharSeq.java]: Yep, I see what you meant that CharSeq is special.

File: javaslang/src/test/java/javaslang/collection/AbstractSeqTest.java (hunk 5 above)
Comment (danieldietrich) [inline javaslang/src/test/java/javaslang/collection/AbstractSeqTest.java]: Let's test both to be robust regarding refactoring. The implementation detail may change.

**Comment by danieldietrich** [comment]
This looks great! I had to think about it a while but finally we have it :-)
Great work (javadoc, junit, the minor fixes et al.)!
